File: backend/kernelCI_app/typeModels/treeCompare.py
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, RootModel

from kernelCI_app.constants.general import DEFAULT_ORIGIN
from kernelCI_app.constants.localization import DocStrings

CompareGroupedStatus = Literal["PASS", "FAIL", "INCONCLUSIVE"]


class CompareStatusCounts(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    pass_count: int = Field(alias="pass", default=0)
    fail_count: int = Field(alias="fail", default=0)
    inconclusive: int = 0


class CompareDelta(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    pass_count: int = Field(alias="pass", default=0)
    fail_count: int = Field(alias="fail", default=0)


class CompareChangeCounts(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    regression: int = 0
    fixed: int = 0
    new_failure: int = Field(alias="newFailure", default=0)
    still_failing: int = Field(alias="stillFailing", default=0)
    new_pass: int = Field(alias="newPass", default=0)
    appeared: int = 0
    disappeared: int = 0


class CompareEntitySummary(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    side_a: CompareStatusCounts = Field(alias="sideA")
    side_b: CompareStatusCounts = Field(alias="sideB")
    delta: CompareDelta
    changes: CompareChangeCounts = Field(default_factory=CompareChangeCounts)


class CompareSummary(BaseModel):
    builds: CompareEntitySummary
    boots: CompareEntitySummary
    tests: CompareEntitySummary


class TreeCompareResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    tree_name: str = Field(alias="treeName")
    branch: str
    git_url: str = Field(alias="gitUrl", default="")
    summary: CompareSummary


class TreeCompareQueryParameters(BaseModel):
    hash_a: str = Field(description="Commit hash for side A")
    hash_b: str = Field(description="Commit hash for side B")
    origin: str = Field(
        default=DEFAULT_ORIGIN,
        description=DocStrings.TREE_QUERY_ORIGIN_DESCRIPTION,
    )


class CompareBuildDiffRow(BaseModel):
    config_name: str
    architecture: str
    compiler: str
    status_a: Optional[CompareGroupedStatus] = None
    status_b: Optional[CompareGroupedStatus] = None
    id_a: Optional[str] = None
    id_b: Optional[str] = None


class TreeCompareBuildsResponse(RootModel[List[CompareBuildDiffRow]]):
    root: List[CompareBuildDiffRow]

File: dashboard/src/types/tree/TreeCompare.ts
import { z } from 'zod';

import { possibleTabs } from '@/types/tree/TreeDetails';

export type CompareStatusCounts = {
  pass: number;
  fail: number;
  inconclusive: number;
};

export type CompareDelta = {
  pass: number;
  fail: number;
};

export const compareChangeTypes = [
  'regression',
  'fixed',
  'newFailure',
  'stillFailing',
  'newPass',
  'appeared',
  'disappeared',
] as const;

export type CompareChangeType = (typeof compareChangeTypes)[number];

/** Row-level classification: the backend counts no category for unchanged pairs. */
export type CompareRowChange = CompareChangeType | 'unchanged';

export type CompareChangeStats = Record<CompareChangeType, number>;

export type CompareEntitySummary = {
  sideA: CompareStatusCounts;
  sideB: CompareStatusCounts;
  delta: CompareDelta;
  changes: CompareChangeStats;
};

export type CompareRevision = {
  hash: string;
  shortHash: string;
  commitName: string;
  date: string;
  tags: string[];
};

export type TreeCompareData = {
  treeName: string;
  branch: string;
  gitUrl: string;
  summary: {
    builds: CompareEntitySummary;
    boots: CompareEntitySummary;
    tests: CompareEntitySummary;
  };
};

/** Side status for an individual compared item. */
export const compareItemStatuses = [
  'PASS',
  'FAIL',
  'INCONCLUSIVE',
  '—',
] as const;
export type CompareItemStatus = (typeof compareItemStatuses)[number];

export type CompareGroupedApiStatus = 'PASS' | 'FAIL' | 'INCONCLUSIVE';

/** Builds detail endpoint: GET .../compare/builds (bare array) */
export type TreeCompareBuildDiffApiRow = {
  config_name: string;
  architecture: string;
  compiler: string;
  status_a: CompareGroupedApiStatus | null;
  status_b: CompareGroupedApiStatus | null;
  id_a: string | null;
  id_b: string | null;
};

/** Boots/tests detail endpoints: GET .../compare/boots|tests (bare array) */
export type TreeCompareTestDiffApiRow = {
  path: string;
  config_name: string;
  architecture: string;
  platform: string;
  status_a: CompareGroupedApiStatus | null;
  status_b: CompareGroupedApiStatus | null;
  id_a: string | null;
  id_b: string | null;
};

export type CompareStatusPair = {
  from: CompareItemStatus;
  to: CompareItemStatus;
};

/** URL/storage token for an empty pair list (show all rows). */
export const compareEmptyStatusPairsToken = 'none';
/** URL token for absent (`—`) so query strings stay ASCII. */
export const compareAbsentStatusToken = 'ABSENT';
export const compareStatusPairStorageKey = 'treeCompare.statusPair';

export const compareDefaultStatusPairParams = ['PASS:FAIL', 'FAIL:PASS'];

type CompareFailureRowBase = {
  id: string;
  change: CompareRowChange;
  sideA: CompareItemStatus;
  sideB: CompareItemStatus;
  idA: string | null;
  idB: string | null;
};

export type CompareBuildFailureRow = CompareFailureRowBase & {
  config: string;
  arch: string;
  compiler: string;
};

export type CompareBootFailureRow = CompareFailureRowBase & {
  path: string;
  config: string;
  arch: string;
  hardware: string;
};

export type CompareTestFailureRow = CompareFailureRowBase & {
  path: string;
  config: string;
  arch: string;
  hardware: string;
};

export type CompareFailureRow =
  | CompareBuildFailureRow
  | CompareBootFailureRow
  | CompareTestFailureRow;

export const compareDefaultValues = {
  hashA: '',
  hashB: '',
  origin: 'maestro',
  currentPageTab: 'global.builds' as const,
};

export const compareSearchSchema = z.object({
  hashA: z.string().catch(''),
  hashB: z.string().catch(''),
  origin: z
    .string()
    .default(compareDefaultValues.origin)
    .catch(compareDefaultValues.origin),
  currentPageTab: z
    .enum(possibleTabs)
    .default(compareDefaultValues.currentPageTab)
    .catch(compareDefaultValues.currentPageTab),
  // Omit = unspecified (hydrate from localStorage, then hardcoded default).
  // `none` is an explicit empty list.
  statusPair: z.preprocess(value => {
    if (value === undefined) {
      return undefined;
    }
    if (value === '') {
      return [compareEmptyStatusPairsToken];
    }
    return Array.isArray(value) ? value : [value];
  }, z.array(z.string()).optional().catch(undefined)),
});

export type CompareSearch = z.infer<typeof compareSearchSchema>;

export const compareRouteName = '/_main/tree/$treeName/$branch/compare';
export const compareNavigateFrom = '/tree/$treeName/$branch/compare';
